samples/bluetooth/peripheral_fast_pair/src/bt_adv_helper.c

```diff
@@ -0,0 +1,218 @@
+ * while the device is actively advertising in Fast Pair not discoverable mode. The value of this
+ * timeout must be lower than CONFIG_BT_RPA_TIMEOUT (3600 seconds) to ensure that RPA rotation will
+ * always trigger update of Account Key Filter advertising data.
+ */
+#define RPA_TIMEOUT_NON_DISCOVERABLE	(13 * SEC_PER_MIN)
+#define RPA_TIMEOUT_OFFSET_MAX		(2 * SEC_PER_MIN)
+#define RPA_TIMEOUT_FAST_PAIR_MAX	(15 * SEC_PER_MIN)
+
+/* The Bluetooth Core Specification v5.2 (Vol. 4, Part E, 7.8.45) allows for time range between
+ * 1 second and 3600 seconds. In case of Fast Pair we should avoid RPA address rotation when device
+ * is Fast Pair discoverable. If not discoverable, the RPA address rotation should be done together
+ * with Fast Pair payload update (responsibility of sample/application).
+ */
+BUILD_ASSERT(CONFIG_BT_RPA_TIMEOUT == 3600);
+
+/* Make sure that RPA rotation will be done together with Fast Pair payload update. */
+BUILD_ASSERT(RPA_TIMEOUT_NON_DISCOVERABLE < CONFIG_BT_RPA_TIMEOUT);
+
+static struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID16_ALL,
+		      BT_UUID_16_ENCODE(BT_UUID_HIDS_VAL),
+		      BT_UUID_16_ENCODE(BT_UUID_BAS_VAL)),
+	/* Empty placeholder for TX power advertising data. */
+	{
+	},
+	/* Empty placeholder for Fast Pair advertising data. */
+	{
+	},
+};
+
+static const struct bt_data sd[] = {
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LEN),
+	BT_DATA_BYTES(BT_DATA_GAP_APPEARANCE,
+		      (CONFIG_BT_DEVICE_APPEARANCE & BIT_MASK(__CHAR_BIT__)),
+		      (CONFIG_BT_DEVICE_APPEARANCE >> __CHAR_BIT__)),
+};
+
+static void rpa_rotate_fn(struct k_work *w);
+static K_WORK_DELAYABLE_DEFINE(rpa_rotate, rpa_rotate_fn);
+
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	struct k_work_sync sync;
+
+	if (!err) {
+		k_work_cancel_delayable_sync(&rpa_rotate, &sync);
+	}
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected        = connected,
+};
+
+static int bt_adv_helper_fast_pair_prepare(struct bt_data *adv_data, bool fp_discoverable)
+{
+	/* Make sure that Fast Pair data was freed and set to NULL to prevent memory leaks. */
+	if (adv_data->data) {
+		k_free((uint8_t *)adv_data->data);
+		adv_data->data = NULL;
+	}
+
+	/* Fast Pair pairing mode must be manually set by the sample. */
+	bt_fast_pair_set_pairing_mode(fp_discoverable);
+
+	size_t buf_size = bt_fast_pair_adv_data_size(fp_discoverable);
+	uint8_t *buf = k_malloc(buf_size);
+
+	if (!buf) {
+		return -ENOMEM;
+	}
+
+	int err = bt_fast_pair_adv_data_fill(adv_data, buf, buf_size, fp_discoverable);
+
+	if (err) {
+		k_free(buf);
+	}
+
+	return err;
+}
+
+static int bt_adv_helper_tx_power_prepare(struct bt_data *adv_data)
+{
+	/* Make sure that TX power data was freed and set to NULL to prevent memory leaks. */
+	if (adv_data->data) {
+		k_free((uint8_t *)adv_data->data);
```

Comment:
The data type in the Zephyr's structure has type: `const uint8_t *`. We need to cast the `const` away to avoid compiler warning when passing it to k_free function. This file manages memory here so we know that the memory was allocated and we must free it.

Reply:
Changed to `void *` to be consistent with `k_free` API